### ISSUE TITLE
Add Serilog to Projects other than C7

### ISSUE THREAD

**Issue by QuintillusCFC** [C7-Game/Prototype#267]
#118 did the hard work for getting a logging framework.  But it's only in the C7 project.  We should have it everywhere.  I'm particularly guilty of adding a lot of Console.WriteLine statements in C7Engine, but this applies to all of them.

This might be a good first issue, as it should be possible to emulate the existing changes in C7.  Though I'm not 100% sure there won't be any complicating issues.

**Comment by WildWeazel** [comment]
This raises a question that I was going to ask for the file reorg: do we need LogManager.ForContext<T>? I think not, as it's just an alias and Serilog.Log is already static. But I ask because other projects need to not use LogManager, because it depends on Godot.

**Comment by QuintillusCFC** [comment]
> This raises a question that I was going to ask for the file reorg: do we need LogManager.ForContext? I think not, as it's just an alias and Serilog.Log is already static. But I ask because other projects need to not use LogManager, because it depends on Godot.

So, looking at #255 and #178 , I think the answer is "it's complicated".  I'm not the .NET expert, but my understanding is the benefit of Log.ForContext (in Serilog) is that it includes the "context", e.g. the type logging the message, so logs can be selectively enabled/disabled by class (e.g. only enable debug messages for components related to the audio if we're debugging an audio bug).  So we do want the Serilog ForContext.

But the LogManager appears to be a helper class that integrates the logs with the Godot sink.  You are right that it can't be used in other projects.

There were already some oddities about logs from projects outside C7, namely that if you run from Godot, they don't show up in the log within the Godot engine window (they do in the command prompt window that starts alongside Godot, however... also this may be completely different on platforms other than Windows).  I think they do show up properly in Rider, but the EngineTests project is currently broken, so I can't compile to test that.  Not sure about VSCode.  But my conclusion is that there may be extra hoops to jump through to get non-Godot-project logs showing up in Godot's console output.

But that's already a hazard; I would put "enable Serilog and its benefits of logging levels and ForContext filtering" above "make the logs show up in the Godot output window", which is merely a convenience.  That's a nice thing to have eventually, but at least for me not a significant pain point, just a minor nuisance.

tl;dr: We should use Serilog's ForContext in the other projects, but they obviously can't be tied to Godot.

**Comment by QuintillusCFC** [comment]
There's another slightly longer term goal that I think should be spun off from this: adding a flat file sink, and preferably making it so there's one place it's configured fro all the projects, rather than C7, C7Engine, C7DataType, etc. etc. having to do completely separate configuration.  The flat file sink enables retrospectively examining what's going on.  Useful for us, but even more useful once we have non-developer users.  I can't tell you how many times someone being able to send me the log has helped figure out what's going on with my editor.  It doesn't always pinpoint the problem (although sometimes it does), but it at least helps figure out a lead.  If I thought, "maybe it's broken at point A", but it keeps right on going to point B, I know my initial theory was (probably) not correct.

Now the thing that confuses me is that LogManager, as it exists now, creates a new LoggerConfiguration, calls CreateLogger, and doesn't do anything with the object returned.  Reading the documentation at https://github.com/serilog/serilog/wiki/Configuration-Basics, I _think_ maybe this configuration becomes the new default for all loggers?  If so, that's similar to what I'm used to with Log4J where you just configure it once at startup, and as long as you don't need to do anything too fancy, all your logs inherit that configuration.  I really like how Log4J does it that way, since it means you can configure your custom log sinks in your application, and all the other code at a lower level inherits that sink without having to know about e.g. Godot, because Log4J just calls the API on its sinks and one of them happens to log to the Godot console, and it magically works.  

In my editor example, my "Civ3SharedComponents" library inherits the configuration from the main "Civ3Editor" Log4J configuration, and it could call classes that are only in "Civ3Editor." If I wanted a custom sink that displayed as an overlay on the map tab, for instance, "Civ3SharedComponents" could still be piped to it with no additional work beyond the UI code for the overlay and a one-line pipe setup.

Hopefully Serilog works similarly, since if it does, we might get our log statements showing up in the Godot console with no additional work beyond the Serilog dependency and a Log.ForContext(MyType).Info("Useful message")

**Comment by QuintillusCFC** [comment]
I have confirmed that in Rider, all logs get piped to the in-IDE console.  Only C7 ones do in Godot.

We do have some formatting problems though, from both Godot and Rider, printing things like:

```
←[38;5;0246m01:30:27.246←[0m←[38;5;0242m [←[0m←[37mDBG←[0m←[38;5;0242m]←[0m←[38;5;0242m ←[0m←[38;5;0246mMainMenu←[0m←[38;5;0242m:←[0m←[38;5;0242m ←[0m←[38;5;0246menter MainMenu._R
eady←[0m←[38;5;0242m

```

Maybe that should be separate as a bug...

Also figured out why my local isn't working, for some reason the net6.0 move broke the Engine tests :(.  Why?  Not sure, I'll have to compare it to the other projects.  That should also probably be spun off...

**Comment by WildWeazel** [comment]
> There's another slightly longer term goal that I think should be spun off from this: adding a flat file sink, and preferably making it so there's one place it's configured fro all the projects, rather than C7, C7Engine, C7DataType, etc. etc. having to do completely separate configuration.

Yes, see #260. We can configure once in LogManager for Godot, or in a test fixture for unit tests, etc, and the projects will only need to get their own loggers.

> Now the thing that confuses me is that LogManager, as it exists now, creates a new LoggerConfiguration, calls CreateLogger, and doesn't do anything with the object returned. Reading the documentation at https://github.com/serilog/serilog/wiki/Configuration-Basics, I _think_ maybe this configuration becomes the new default for all loggers?

That's my expectation, and seems to be the case now that you've added loggers to C7Engine. For example if I change the MinimumLevel in LogManager it applies to logs from both projects.
More specifically, Log.Logger is a static instance described as "the globally-shared logger" and subsequent calls to Log.ForContext are handled by this instance.

> Hopefully Serilog works similarly, since if it does, we might get our log statements showing up in the Godot console with no additional work beyond the Serilog dependency and a Log.ForContext(MyType).Info("Useful message")

With your changes I'm seeing C7Engine Serilog statements in both the Godot output tab and the terminal. We can have the loggers pipe to as many sinks as we want (Godot, file, in-game...) and configure them all in once place.

**Comment by QuintillusCFC** [comment]
^That all makes sense, and I wound up answering some of my own questions some time between 12 and 2 days ago.

I'm leaving this as "In Progress" for now since I've only added it to the Engine; there are still other places to add it (and logs to convert in the main C7 project).  I'll unassign myself so if someone else want to start on the rest of them, they can.  I'll probably swing back around to it once I've wrapped up my other in-progress items, or earlier if they'd benefit from the rest of this being done first.
